SystemManager/Src/SystemManager.cpp

```diff
@@ -10,37 +10,56 @@
 #include "drivers_config.hpp"
 #include "independent_watchdog.h"
 #include "tim.h"
+#include "CommonDataTypes.hpp"
+#include "config.hpp"
+#include "cmsis_os.h"
 
 SystemManager::SystemManager():
     rcController_(&huart2),
     throttleMotorChannel_(&htim2, TIM_CHANNEL_1),
     yawMotorChannel_(&htim2, TIM_CHANNEL_2),
     rollMotorChannel_(&htim2, TIM_CHANNEL_3),
     pitchMotorChannel_(&htim2, TIM_CHANNEL_4),
-    invertedRollMotorChannel_(&htim3, TIM_CHANNEL_1),
-    watchdog_(&hiwdg)
-{}
+    watchdog_(&hiwdg),
+    am_instance_(config::flightmodes[config::DEFAULT_FLIGHTMODE].flightmodeConstructor())
+{
+    xTaskCreate(runAM, "AM Thread", 400U, (void*)&am_instance_, osPriorityNormal, &AM_handle_);
+}
+
+SystemManager::~SystemManager() {
+    vTaskDelete(AM_handle_);
+}
 
-SystemManager::~SystemManager() {}
+void SystemManager::runAM(void* pvParameters) {
+    AM::AttitudeManager* am_instance = (AM::AttitudeManager *) pvParameters;
+    TickType_t xLastWakeTime;
+    xLastWakeTime = xTaskGetTickCount();
+    for(;;) {
+        am_instance->runControlLoopIteration(AM::AttitudeManager::getControlInputs());
+        vTaskDelayUntil(&xLastWakeTime, AM_PERIOD_MS);
+    }
+}
 
 void SystemManager::flyManually() {
     for(;;){
         this->rcInputs_ = rcController_.GetRCControl();
         if (this->rcInputs_.isDataNew) watchdog_.refreshWatchdog();
+        
+        AM::AttitudeManagerInput am_input = {
+            .roll = 0,
+            .pitch = 0,
+            .yaw = 0,
+            .throttle = 0
+        };
 
-        if(this->rcInputs_.arm >= (SBUS_MAX/2)) {
-            this->throttleMotorChannel_.set(rcInputs_.throttle);
-            this->yawMotorChannel_.set(rcInputs_.yaw);
-            this->rollMotorChannel_.set(rcInputs_.roll);
-            this->pitchMotorChannel_.set(rcInputs_.pitch);
-            this->invertedRollMotorChannel_.set(SBUS_MAX - rcInputs_.roll);
-        }
-        else{
-            this->throttleMotorChannel_.set(0);
-            this->yawMotorChannel_.set(0);
-            this->rollMotorChannel_.set(0);
-            this->pitchMotorChannel_.set(0);
-            this->invertedRollMotorChannel_.set(0);
+        if (this->rcInputs_.arm >= (SBUS_MAX/2)) {
+            am_input.roll = rcInputs_.roll;
+            am_input.pitch = rcInputs_.pitch;
+            am_input.yaw = rcInputs_.yaw;
+            am_input.throttle = rcInputs_.throttle;
+            AM::AttitudeManager::setControlInputs(am_input);
+        } else {
+            AM::AttitudeManager::setControlInputs(am_input);
```

Comment:
Do you think we should add arm and disarm methods directly into AM, so as to ensure that we can't accidentally write some value to the motors if we are in a flight mode that does not just directly map the controller inputs?

Reply:
Are you saying we should add arm/disarm logic to `AM::setControlInputs()`? I'm guessing you want to disarm based on some flightmode property.